Parallel testing + unveiled logic flaws

KtorZ opened the pull request:
- :round_pushpin: **Run hydra-node tests in parallel.**
  
- :round_pushpin: **Fix race-condition during snapshot creation.**
    Unfortunately. I did change a bit the behavior of the snapshot creation so that snapshots are created right after every transaction. Yet, it's done in such a way that non-leaders will however keep accumulating in flights transactions and should be able to make snapshots with more than one txs when their turn come. Thus, I do not expect to see much performance hit on the benchmarks, but haven't run them yet.

- :round_pushpin: **re-generate JSON golden tests after removing the 'DoSnapshot' instruction.**
  
- :round_pushpin: **Keep track of programmed events**
    So that 'isEmpty' for the event queue return true only if there's no event in flights (and the queue is empty!).
  This actually caught quite a few bugs in our logic which was firing
  multiple 'DoSnapshots' and ignoring some events... (hence the previous commits)

KtorZ committed:
Run hydra-node tests in parallel.

KtorZ committed:
Fix race-condition during snapshot creation.

  Unfortunately. I did change a bit the behavior of the snapshot creation so that snapshots are created right after every transaction. Yet, it's done in such a way that non-leaders will however keep accumulating in flights transactions and should be able to make snapshots with more than one txs when their turn come. Thus, I do not expect to see much performance hit on the benchmarks, but haven't run them yet.

KtorZ committed:
re-generate JSON golden tests after removing the 'DoSnapshot' instruction.

KtorZ committed:
Keep track of programmed events

  So that 'isEmpty' for the event queue return true only if there's no event in flights (and the queue is empty!).
  This actually caught quite a few bugs in our logic which was firing
  multiple 'DoSnapshots' and ignoring some events... (hence the previous commits)

ch1bo commented on hydra-node/src/Hydra/Node.hs:
```diff
@@ -167,7 +173,7 @@ processEffect HydraNode{hn, oc, server, eq, env = Environment{party}} tracer e =
     NetworkEffect msg -> broadcast hn msg >> putEvent eq (NetworkEvent msg)
     OnChainEffect tx -> postTx oc tx
-    Delay after event -> void . async $ threadDelay after >> putEvent eq event
+    Delay after event -> putEventAfter eq after event
```
This is nice. I wonder why we even have the `Delay` now. What about

```haskell
processNextEvent HydraNode{hh, env} e =
  modifyHeadState hh $ \s ->
    case Logic.update env (ledger hh) s e of
      NewState s' effects -> (NewState s' effects, s')
      Error err -> (Error err, s)
      Wait -> (Wait, s)
```

```haskell
stepHydraNode tracer node@HydraNode{eq, env = Environment{party}} = do
  e <- nextEvent eq
  traceWith tracer $ ProcessingEvent party e
  atomically (processNextEvent node e) >>= \case
    NewState _ effs -> forM_ effs (processEffect node tracer) >> traceWith tracer (ProcessedEvent party e)
    Error err -> traceWith tracer (ErrorHandlingEvent party e err)
    Wait -> putEventAfter eq 0.1 e
```

And remove this line as well as the `Delay` effect altoghether? :slightly_smiling_face:  :tada:

KtorZ replied:
I actually got confused for a bit initially. Seeing that we have `Wait` as an outcome, but a `Delay` effect. Yet, we only use the `Delay` effect to re-enqueue events which is exactly the semantic of `Wait`! Hence, I guess you're right :)

ch1bo replied:
Ah yeah.. there is the ShouldPostFanout as well. Fine, maybe we'll get rid of some of these things later.. or not.

ch1bo commented on hydra-node/hydra-node.cabal:
```diff
@@ -176,6 +176,7 @@ executable mock-chain
 
 test-suite tests
   import:             project-config
+  ghc-options:        -threaded -rtsopts -with-rtsopts=-N2
```
Isn't `-N` without `2` the default and uses as many cores as available? (unless overridden when calling)

KtorZ replied:
With `-N`, my CPU goes above 900% usage and my fans get crazy for a sec. Plus, because most of our tests are quite short, we don't get much benefits from _too much parallelism_ as the scheduler spend more time managing threads than doing actual work :sweat_smile: ... 

I've tried some parameters, and seems like 2 is nice enough, provides speed up but keeps the laptop cool and works against climate change :fire:

ch1bo replied:
As we are running tests through cabal, there might be a way to have per-user configuration. I for my part would like to benefit from the 24 threads of my cpu 😅

ch1bo commented on hydra-node/src/Hydra/HeadLogic.hs:
```diff
@@ -224,28 +223,44 @@ update Environment{party, signingKey, otherParties, snapshotStrategy} ledger st
+                    (newSeenSnapshot, effects) = case seenSnapshot of
+                      NoSeenSnapshot
+                        | shouldDoSnapshot confirmedSnapshot ->
+                          ( RequestedSnapshot
```
We do need `RequestedSnapshot` to avoid emitting multiple `ReqSn` when we have multiple `ReqTx` in queue.

KtorZ replied:
As @ch1bo pointed out indeed.

KtorZ commented on hydra-node/src/Hydra/Node.hs:
```diff
@@ -177,18 +183,33 @@ processEffect HydraNode{hn, oc, server, eq, env = Environment{party}} tracer e =
+          void . async $ do
+            threadDelay delay
+            atomically $ do
+              modifyTVar' numThreads (\n -> n - 1)
```
:clap: yes!

KtorZ commented on hydra-node/test/Hydra/BehaviorSpec.hs:
```diff
@@ -48,294 +48,301 @@ import Test.Aeson.GenericSpecs (roundtripAndGoldenSpecs)
+            FailureDeadlock _ -> True
+            _ -> False
+
+    parallel $
```
The docs says `parallel marks all spec items of the given spec to be safe for parallel evaluation.`. Now, it's a bit unclear what an _items_ of a spec is, and if it's only the direct descendant or also the nested children.

KtorZ committed:
Acknowledge PR reviews after introducing parallel testing

  - Use `pred` and `succ` instead of lambdas with `modifyMVar`.
  - Use `rtsopts=-N` instead of a fixed number of cores.
  - Use a single `parallel` per spec tree, it's recursive.
